sync/pending_polling.go

```diff
@@ -0,0 +1,673 @@
+			targetPreConfirmedNum, stagedPreLatest = s.handlePreLatest(
+				pl,
+				targetPreConfirmedNum,
+				stagedPreLatest,
+			)
+		case pc := <-preConfirmedCh:
+			s.handlePreConfirmed(pc, targetPreConfirmedNum, stagedPreLatest)
+		}
+	}
+}
+
+// pollPendingData runs synchronisation in two phases, switching when protocol >= v0.14.0:
+//  1. Pending phase: store empty pending baselines and real pending blocks.
+//  2. Pre_confirmed phase: manage baselines and pre_latest attachments and poll pre_confirmed.
+func (s *Synchronizer) pollPendingData(ctx context.Context) {
+	if s.pendingPollInterval == 0 || s.preConfirmedPollInterval == 0 {
+		s.log.Infow("Pending data polling is disabled")
+		return
+	}
+
+	headsSub := s.newHeads.SubscribeKeepLast()
+	defer headsSub.Unsubscribe()
+
+	// Phase 1: pending path
+	switched := s.runPendingPhase(ctx, headsSub)
+	if !switched || ctx.Err() != nil {
+		return
+	}
+
+	// Phase 2: pre_confirmed path
+	s.runPreConfirmedPhase(ctx, headsSub)
+}
+
+// StorePending stores a pending block given that it is for the next height
+func (s *Synchronizer) StorePending(p *core.Pending) (bool, error) {
+	err := blockchain.CheckBlockVersion(p.Block.ProtocolVersion)
+	if err != nil {
+		return false, err
+	}
+
+	expectedParentHash := new(felt.Felt)
+	h, err := s.blockchain.HeadsHeader()
+	if err != nil && !errors.Is(err, db.ErrKeyNotFound) {
+		return false, err
+	} else if err == nil {
+		expectedParentHash = h.Hash
+	}
+
+	if !expectedParentHash.Equal(p.Block.ParentHash) {
+		return false, fmt.Errorf("store pending: %w", blockchain.ErrParentDoesNotMatchHead)
+	}
+
+	if existingPending, err := s.PendingData(); err == nil {
+		if existingPending.GetBlock().TransactionCount >= p.Block.TransactionCount {
+			// ignore the incoming pending if it has fewer transactions than the one we already have
+			return false, nil
+		}
+	} else if !errors.Is(err, ErrPendingBlockNotFound) {
+		return false, err
+	}
+
+	s.pendingData.Store(utils.HeapPtr[core.PendingData](p))
+
+	return true, nil
+}
+
+// UpdatePreLatestAttachment updates (or clears) the PreLatest attachment of the currently stored
+// pre_confirmed at the given blockNumber by atomically swapping the stored pointer.
+// Returns true if the store was updated, false if no matching pre_confirmed is stored
+// or the attachment was already equal.
+func (s *Synchronizer) UpdatePreLatestAttachment(blockNumber uint64, preLatest *core.PreLatest) bool {
+	curPtr := s.pendingData.Load()
+	if curPtr == nil || *curPtr == nil {
+		return false
+	}
+
+	cur := *curPtr
+	pc, ok := cur.(*core.PreConfirmed)
+	if !ok {
+		// A pending is stored or another type; nothing to do.
+		return false
+	}
+	if pc.Block == nil || pc.Block.Number != blockNumber {
+		// Different height stored; do not touch.
+		return false
+	}
+
+	if pc.PreLatest == preLatest {
+		// No change.
+		return false
+	}
+
+	// Clone and update attachment on the clone.
+	next := pc.Copy()
+	next.WithPreLatest(preLatest)
```

Comment:
Also the nil safety is redundant, since `next.WithPrelatest(...)` will fail if `next` is nil. I believe all of this should be wrapped inside an `if` in that case (and if that was the intention).